File: 2018/10/201810.jl
using DelimitedFiles

function read_data(inpath::String)::Tuple{Array{Int,2},Array{Int,2}}
    r = []
    for line in eachline(inpath)
        m = match(r"position=< *(-?[0-9]+), *(-?[0-9]+)> velocity=< *(-?[0-9]+), *(-?[0-9]+)>", line)
        isnothing(m) && continue
        push!(r, [parse(Int, c) for c in m.captures])
    end
    pos = [r[i][j] for i = 1:length(r), j = 1:2]
    vel = [r[i][j] for i = 1:length(r), j = 3:4]
    (pos, vel)
end

# minimum spread
function metric(pos::Array{Int,2})::Int
    (mnx,mxx) = extrema(pos[:,1])
    (mny,mxy) = extrema(pos[:,2])
    -(mxx-mnx) * (mxy-mny)
end

function sim(pos::Array{Int,2}, vel::Array{Int,2}, n::Int)::Int
    lastmt = typemin(Int)
    for i = 1:n
        pos .+= vel
        mt = metric(pos)
        if mt <= lastmt
            pos .-= vel
            j = i - 1
            writedlm("snap.$j.csv", pos, ",")
            println(i, " ", mt)
            break
        end
        lastmt = mt
    end
    lastmt
end

function main()
    #(pos,vel) = read_data("tinput.txt")
    (pos,vel) = read_data("input.txt")
    res = sim(pos, vel, 20000)
    println("PART1: ", res)
end

main()

File: 2018/10/201810.py
import numpy as np
import re

def read_data(inpath):
    pos = []
    vel = []
    for line in open(inpath):
        m = re.match(r"position=< *(-?[0-9]+), *(-?[0-9]+)> velocity=< *(-?[0-9]+), *(-?[0-9]+)>", line)
        if m:
            ms = [int(g) for g in m.groups()]
            pos.append([ms[0], ms[1]])
            vel.append([ms[2], ms[3]])
    return (np.array(pos), np.array(vel))

def metric(pos):
    mnx = np.amin(pos[:,0])
    mxx = np.amax(pos[:,0])
    mny = np.amin(pos[:,1])
    mxy = np.amax(pos[:,1])
    return -(mxx - mnx) * (mxy - mny)
    
def sim(pos, vel, n):
    lastmt = -999999999999999
    for i in range(n):
        pos += vel
        mt = metric(pos)
        if mt <= lastmt:
            pos -= vel
            np.savetxt("snappy.{}.csv".format(i-1), pos, fmt="%d", delimiter=',')
            break
        lastmt = mt
    return i-1

def main():
    pos,vel = read_data("input.txt")
    r = sim(pos, vel, 20000)
    print(r)

main()

